### DISCUSSION TITLE
Question about code

### DISCUSSION THREAD

**Discussion by Hang95** [relab/hotstuff#35] [Q&A] [answered]
Hi,sorry to bother you. I'm new to Golang and try to learn the language through your code. I notice that when using values in a struct, like
```
type Block struct {
	hash     Hash
	parent   Hash
	proposer hotstuff.ID
	cmd      Command
	cert     QuorumCert
	view     View
}
```
the code  defined a method 
```
func (b *Block) Hash() Hash {
	return b.hash
}
```
to return it instead of directly using `Block.hash`. Why to use it this way?
Thank you so much.

**Comment by johningve** [comment, marked as answer]
Hi!

There are a few reasons to do this in Go:

1. To make a field immutable (at least to other packages)
2. To be able to use it in an interface

As for `Block.Hash()`, the main reason is to make the hash field immutable. Also, the `Hash()` method was previously used to calculate the hash the first time it was called, but I moved that to `NewBlock`.

**Reply by Hang95** [reply]
Thank you!
